src/platform/GenericOTARequestorDriver.cpp

```diff
@@ -46,29 +57,77 @@ uint16_t GenericOTARequestorDriver::GetMaxDownloadBlockSize()
     return 1024;
 }
 
-void GenericOTARequestorDriver::HandleError(UpdateFailureState state, CHIP_ERROR error)
+void StartDelayTimerHandler(System::Layer * systemLayer, void * appState)
+{
+    static_cast<GenericOTARequestorDriver *>(appState)->DriverSendQuery();
+}
+
+bool ProviderLocationsEqual(const ProviderLocation::Type & a, const ProviderLocation::Type & b)
 {
-    // TODO: Schedule the next QueryImage
+    if ((a.fabricIndex == b.fabricIndex) && (a.providerNodeID == b.providerNodeID) && (a.endpoint == b.endpoint))
+    {
+        return true;
+    }
+    else
+    {
+        return false;
+    }
+}
+
+void GenericOTARequestorDriver::HandleError(UpdateFailureState state, CHIP_ERROR error) {}
+
+void GenericOTARequestorDriver::HandleIdleState()
+{
+    // Default profivers timer runs if and only if the OTARequestor's update state is kIdle.
+    // Must (re)start the timer every time we enter the kIdle state
+    StartDefaultProviderTimer();
 }
 
 void GenericOTARequestorDriver::UpdateAvailable(const UpdateDescription & update, System::Clock::Seconds32 delay)
 {
+    // IMPLEMENTATION CHOICE:
+    // This implementation unconditionally downloads an available update
+
     VerifyOrDie(mRequestor != nullptr);
-    ScheduleDelayedAction(UpdateFailureState::kDownloading, delay,
-                          [](System::Layer *, void * context) { ToDriver(context)->mRequestor->DownloadUpdate(); });
+    ScheduleDelayedAction(
+        delay, [](System::Layer *, void * context) { ToDriver(context)->mRequestor->DownloadUpdate(); }, this);
 }
 
 void GenericOTARequestorDriver::UpdateNotFound(UpdateNotFoundReason reason, System::Clock::Seconds32 delay)
 {
     VerifyOrDie(mRequestor != nullptr);
 
+    ProviderLocation::Type providerLocation;
+
+    switch (reason)
+    {
+    case UpdateNotFoundReason::UpToDate:
+        return;
+        break;
+    case UpdateNotFoundReason::Busy:
+        // This will schedule a query with the same provider
+        break;
+    case UpdateNotFoundReason::ConnectionFailed:
```

Comment:
This failure doesn't feel like it should fit in with the UpdateNotFound list of reasons. Perhaps, we should combine `UpdateNotFound` with `HandleError` as one more generic function. The param that is passed in would then be used similarly here as `reason` and determine what to do next in the driver.

Reply:
I see your point, yet having an all-purpose `HandleError` doesn't seem too clean either -- we'd have to pass additional parameters to it essentially identifying separately each place of the code it's called from (what state we are in? what kind of error occurred?) -- because actions would be different. I'd rather not deal with it in this PR.

Reply:
Agreed that it doesn't have to be handled in this PR but I think having one API to call makes it less confusing for the OTA Requestor core side to determine which API it should call for the next action to occur.